### PR TITLE
Imgui multi canvas

### PR DESCRIPTION
**Pull request by kushalkolar**
Exposes the context of `ImguiRenderer` so that it can be used for context switching. As far as I understand, this is required for rendering different UIs to multiple different canvases (added an example). Not sure if there's a better way to do this, welcome to suggestions! @panxinmiao 

tested in glfw & qt

### CODE REVIEW

**Comment by panxinmiao** [comment]
I didn't consider the situation of multiple` ImguiRenderer` instances. 😅 

LGTM.

File: examples/imgui_multi_canvas.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    imgui_renderer3.render(update_gui3())
+
+    # done! request draw
+    canvas.request_draw()
+    canvas2.request_draw()
+    canvas3.request_draw()
+
+
+if __name__ == "__main__":
+    canvas.request_draw(draw_frame)
```
Comment (kushalkolar) [inline examples/imgui_multi_canvas.py:97]: this is maybe a bit weird, I think this should be separated into 3 different `draw_frame()` functions for each canvas?

**Comment by kushalkolar** [comment]
> I didn't consider the situation of multiple` ImguiRenderer` instances. 😅

No worries, your work on implementing imgui integration has been amazing! :smile:

File: examples/imgui_multi_canvas.py (hunk 1 above)
Comment (almarklein) [reply examples/imgui_multi_canvas.py:97]: Yes, I think that would be the correct way. Although there's one event-loop so the current implementation should also be fine.

File: examples/imgui_multi_canvas.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    imgui.end_frame()
+    imgui.render()
+
+    return imgui.get_draw_data()
+
+
+def draw_frame():
+    # set corresponding imgui context before rendering
+    imgui.set_current_context(imgui_renderer.imgui_context)
+    imgui_renderer.render(update_gui())
```
Comment (Korijn) [inline examples/imgui_multi_canvas.py:82]: Why did you decide not to call set current context within the render function? Seems like if it's always necessary you could just do it there?

Comment (panxinmiao) [reply examples/imgui_multi_canvas.py:82]: It's not the `render` function that needs it; the `render` method only requires `draw_data`. However, to obtain `draw_data`, we need to switch to the correct context before updating the GUI code. This means that setting the context should be done in the `update_gui()` method. The user needs to build or update the GUI in the correct imgui context environment.

If we want to handle this within the `render` method, we need to modify the API by adding a property to the `ImguiRenderer` as a hook function for updating the GUI. Then, in the `render` function, we would do something like this:

```python
imgui.set_current_context(...)
draw_data = self.update_gui_callback()  # Call the hook method
# Do render with draw_data
```

Comment (kushalkolar) [reply examples/imgui_multi_canvas.py]: > If we want to handle this within the `render` method, we need to modify the API by adding a property to the `ImguiRenderer` as a hook function for updating the GUI. Then, in the `render` function, we would do something like this:
> 
> ```python
> imgui.set_current_context(...)
> draw_data = self.update_gui_callback()  # Call the hook method
> # Do render with draw_data
> ```

I think I like this idea! I updated the PR to implement this. Users do not have to be aware of context switching.

Comment (almarklein) [reply examples/imgui_multi_canvas.py]: The other imgui examples will need to be updated to use the new API, right?

Comment (kushalkolar) [reply examples/imgui_multi_canvas.py]: Yup, I wanted your thoughts on the changes before modifying the other examples.

Comment (kushalkolar) [reply examples/imgui_multi_canvas.py]: I went ahead and updated the other examples,if anyone (especially @panxinmiao ) has more suggestions I can incorporate them :)

Comment (panxinmiao) [reply examples/imgui_multi_canvas.py]: Thanks for the updates! 
LGTM :)

**Review by panxinmiao — APPROVED** [review]

**Review by Korijn — APPROVED** [review]

**Pull request merged**